src/maker.rs

```diff
@@ -0,0 +1,394 @@
+    ) -> Self {
+        Maker {
+            btc_balance,
+            dai_balance,
+            btc_fee,
+            dai_fee,
+            btc_reserved_funds: Default::default(),
+            dai_reserved_funds: Default::default(),
+            btc_max_sell_amount,
+            dai_max_sell_amount,
+            mid_market_rate,
+            spread,
+            ongoing_takers: Default::default(),
+        }
+    }
+
+    pub fn update_rate(&mut self, mid_market_rate: ()) {}
+
+    pub fn update_bitcoin_balance(&mut self, balance: bitcoin::Amount) {
+        self.btc_balance = balance;
+    }
+
+    pub fn update_dai_balance(&mut self, balance: dai::Amount) {
+        self.dai_balance = balance;
+    }
+
+    pub fn track_failed_rate(&mut self, error: anyhow::Error) {}
+
+    pub fn track_failed_balance_update(&mut self, error: anyhow::Error) {}
+
+    pub fn next_sell_order(&self) -> anyhow::Result<BtcDaiOrder> {
+        BtcDaiOrder::new_sell(
+            self.btc_balance,
+            self.btc_fee,
+            self.btc_reserved_funds,
+            self.btc_max_sell_amount,
+            self.mid_market_rate.value,
+            self.spread,
+        )
+    }
+
+    pub fn next_buy_order(&self) -> anyhow::Result<BtcDaiOrder> {
+        BtcDaiOrder::new_buy(
+            self.dai_balance.clone(),
+            self.dai_fee.clone(),
+            self.dai_reserved_funds.clone(),
+            self.dai_max_sell_amount.clone(),
+            self.mid_market_rate.value,
+            self.spread,
+        )
+    }
+
+    /// Decide whether we should proceed with order,
+    /// Confirm with the order book
+    /// Re & take & reserve
+    pub fn react_to_taken_order(&mut self, order: Order) -> anyhow::Result<Reaction> {
+        if self.ongoing_takers.cannot_trade_with_taker(&order.taker) {
+            return Ok(Reaction::CannotTradeWithTaker);
+        }
+
+        let current_profitable_rate = self
+            .spread
+            .apply(self.mid_market_rate.value, order.inner.position)?;
+        let order_rate = Rate::try_from(order.inner.clone())?;
+        let next_order = match order.clone().into() {
+            order
+            @
+            BtcDaiOrder {
+                position: Position::Buy,
+                ..
+            } => {
+                // We are buying BTC for DAI
+                // Given an order rate of: 1:9000
+                // It is NOT profitable to buy, if the current rate is greater than the order rate.
+                // 1:8800 -> We give less DAI for getting BTC -> Good.
+                // 1:9200 -> We have to give more DAI for getting BTC -> Sucks.
+                if order_rate > current_profitable_rate {
+                    return Ok(Reaction::RateSucks);
+                }
+
+                let updated_dai_reserved_funds = self.dai_reserved_funds.clone() + order.quote;
+                if updated_dai_reserved_funds > self.dai_balance {
+                    return Ok(Reaction::InsufficientFunds);
```

Comment:
As per my other comment, this should be a `Err`. We are interrupting the process with returning an `Result::Ok`, it's weird.